src/main/java/io/vertx/core/eventbus/impl/ConsumerStream.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.core.eventbus.impl;
+
+import io.vertx.codegen.annotations.Nullable;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.eventbus.DeliveryOptions;
+import io.vertx.core.eventbus.EventBus;
+import io.vertx.core.eventbus.Message;
+import io.vertx.core.eventbus.MessageConsumer;
+import io.vertx.core.streams.ReadStream;
+import io.vertx.core.streams.impl.InboundBuffer;
+
+import java.util.UUID;
+
+import static io.vertx.core.eventbus.impl.ProducerStream.STREAM_ADDRESS;
+import static io.vertx.core.eventbus.impl.ProducerStream.STREAM_CREDITS;
+
+public class ConsumerStream<T> implements ReadStream<T>, Handler<Message<T>>  {
+
+  public static <T> void bind(Vertx vertx, String address, Handler<ReadStream<T>> handler, Handler<AsyncResult<Void>> completionHandler) {
+    Context context = vertx.getOrCreateContext();
+    Handler<ReadStream<T>> h = handler;
+    EventBus bus = context.owner().eventBus();
+    bus.consumer(address, msg -> {
+      String remoteAddress = msg.headers().get(STREAM_ADDRESS);
+      if (remoteAddress == null) {
+        // No back pressure (not supported)
+        msg.fail(0, "No credits address");
+        return;
+      }
+      String localAddress = UUID.randomUUID().toString();
+      ConsumerStream<T> stream = new ConsumerStream<>(bus, context);
```

Comment:
This would create new ConsumerStream on each message, should it be put before `bus.consumer()` ?